src/ralph/models/converter.py

```diff
@@ -1,65 +1,82 @@
 
     def __init__(self):
-        """Initializes BaseConverter."""
+        """Initializes BaseConversionSet"""
 
-        self.field_array = []
-        self.prepare_converter(self.field_array, self.get_conversion_set())
+        self._conversion_items = self._get_conversion_items()
 
     @abstractmethod
-    def get_conversion_set(self):
-        """Returns a conversion set used for conversion.
-
-        Each element in the conversion set should be a tuple of length 3.
-        - The first element of each tuple is a destination path string.
-        - The second element can be a source path string or `None`.
-        - The third element defines a transformer function with one argument.
-        """
-
-    def convert(self, event: dict, uuid: str):
-        """Converts and serializes the event.
-
-        Args:
-            event (dict): the event to convert and serialize.
-            uuid (str): the UUID string associated with the event.
-        Returns:
-            (model): converted pydantic model.
-        """
-
-        converted_event = {}
-        for destination_path, source_path, transformer in self.field_array:
-            field = get_dict_value_from_path(event, source_path)
-            transformed_field = transformer(field)
-            if transformed_field not in [None, "", {}]:
-                set_dict_value_from_path(
-                    converted_event, destination_path, transformed_field
-                )
-
-        converted_event["id"] = uuid
-        return self.__model__(**converted_event)
-
-    @staticmethod
-    def prepare_converter(field_array, conversion_set):
-        """Fills the converters field_array using the conversion_set."""
-
-        for field in conversion_set:
-            destination_path = field[0].split(MODEL_PATH_SEPARATOR)
-            source_path = field[1].split(MODEL_PATH_SEPARATOR) if field[1] else None
-            transformer = field[2]
-            field_array.append((destination_path, source_path, transformer))
+    def _get_conversion_items(self) -> set[ConversionItem]:
+        """Returns a set of ConversionItems used for conversion."""
+
+    def __iter__(self):
+        return iter(self._conversion_items)
+
+
+def convert(event: dict, conversion_set: BaseConversionSet, uuid: str) -> BaseModel:
+    """Converts the event using the provided converter and UUID."""
+
+    converted_event = {}
+    for conversion_item in conversion_set:
+        transformed_field = conversion_item.get_value(event)
+        if transformed_field not in [None, "", {}]:
+            set_dict_value_from_path(
+                converted_event, conversion_item.dest, transformed_field
+            )
+    converted_event["id"] = uuid
+    return conversion_set.__dest__(**converted_event)
+
+
+def convert_raw(
+    raw_event: str, converter: BaseConversionSet, uuid_namespace: UUID
+) -> BaseModel:
+    """Converts the raw event string using the provided converter."""
+
+    event = json.loads(raw_event)
```

Comment:
Maybe we should secure this and raise a custom exception if the raw event is poorly formatted?

Reply:
Good Idea) Thanks) updated) Now it's raising a `BadFormatException`